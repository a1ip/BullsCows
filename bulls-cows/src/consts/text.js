export const STAGES = {
    guess: 'guess',
    lose: 'lose',
    win: 'win',
    start: 'start'
}

export const FINAL = {
    lose: 'К сожалению, угадать не удалось!',
    win: 'Ура! Код угадан!',
    attempts: 'Итоговое количество попыток: '
}