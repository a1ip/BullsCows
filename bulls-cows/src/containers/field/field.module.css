.field {
    min-height: 400px;
    margin-top: 64px;
    padding: 16px;
}
/* custom scroll */

/* total width */
body::-webkit-scrollbar {
    background-color: transparent;
    width:8px
}

/* background of the scrollbar except button or resizer */
body::-webkit-scrollbar-track {
    background-color: transparent;
}

/* scrollbar itself */
body::-webkit-scrollbar-thumb {
    background-color:#babac0;
    border-radius:8px;
}

/* set button(top and bottom of the scrollbar) */
body::-webkit-scrollbar-button {display:none}