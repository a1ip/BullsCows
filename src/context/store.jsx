import React, {createContext, useReducer} from 'react'
import {STAGES} from "../consts/text"
import {shuffleArray} from "../utils/utils"
const base = [1,2,3,4,5,6,7,8,9,0]

const initialState = {stage: STAGES.start, code: null}
const store = createContext(initialState)
const { Provider } = store

const StateProvider = ( { children } ) => {
    const [state, dispatch] = useReducer((state, action) => {
        switch(action.type) {
            case 'guess':
                shuffleArray(base)
                const code = base.slice(0,4)
                return {...state, stage: STAGES.guess, attempts: 0, code: code}
            case 'win':
                return {...state, stage: STAGES.win, attempts: action.attempts}
            case 'lose':
                return {...state, stage: STAGES.lose, attempts: action.attempts}
            default:
                return initialState
        }
    }, initialState)

    return <Provider value={{ state, dispatch }}>{children}</Provider>
}

export { store, StateProvider }