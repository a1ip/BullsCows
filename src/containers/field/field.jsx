import React, {useState, useContext, useEffect} from 'react'
import style from './field.module.css'
import {Row} from "../../components/row-guess/row"
import {InputGuessField} from "../input-guess-field/input-guess-field"
import {Counter} from "../../components/bulls-cows-counter/counter"
import {Blocks} from "../../components/code-blocks/code-blocks"
import {compareArrays} from "../../utils/utils"
import {STAGES} from "../../consts/text"
import {store} from "../../context/store"

export const Field = ({code, tries}) => {
    const globalState = useContext(store);
    const { dispatch } = globalState;

    const [attempts, setAttempt] = useState([])
    const [currentStage, setStage] = useState(globalState.state.stage)

    const originalCode = code.join('')
    const checkValue = (userCode) => {
        setAttempt([userCode, ...attempts,])
    }
    useEffect(() => {
        setAttempt([])
        setStage(globalState.state.stage)
    }, [code])

    const userCheckedInput = (initialCode, userCode) => {
        const checkedInput = {}
        if (userCode === initialCode) {
            checkedInput.bulls = 4
            checkedInput.cows = 0
            checkedInput.stage = STAGES.win

            if (currentStage !== STAGES.win) {
                setStage(STAGES.win)
                dispatch({ type: STAGES.win, attempts: attempts.length })
            }

        } else {
            const checkedResult = compareArrays([...initialCode], [...userCode])
            checkedInput.bulls = checkedResult.exact
            checkedInput.cows = checkedResult.values
            checkedInput.stage = STAGES.guess
        }

        if (tries > 0 && tries === attempts.length && userCode === attempts[0] && checkedInput.bulls < 4) {
            checkedInput.stage = STAGES.lose

            if (currentStage !== STAGES.lose) {
                setStage(STAGES.lose)
                dispatch({ type: STAGES.lose, attempts: attempts.length })
            }
        }

        return checkedInput
    }

    return (
        <div className={style.field}>
            {(currentStage === STAGES.guess || attempts.length < 1) &&
                <Row>
                    <InputGuessField checkValue={checkValue}/>
                </Row>
            }
            {attempts.length > 0 && attempts.map(item => {
                const result = userCheckedInput(originalCode, item)

                return (
                    <Row>
                        <Blocks stage={result.stage} digits={[...item]}/>
                        <Counter bulls={result.bulls} cows={result.cows} total={item.length}/>
                    </Row>)
            })}
        </div>
    )

}