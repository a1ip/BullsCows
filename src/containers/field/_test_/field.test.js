import React from "react"
import renderer from 'react-test-renderer'
import {store} from "../../../context/store"


import {Field} from "../field"
import {STAGES} from "../../../consts/text"

describe('Field container', () => {

    const FieldTest = renderer
        .create(
            <store.Provider value={{state: {stage: STAGES.guess, code: [2, 3, 4, 5]}}}>
                <Field code={[3, 2, 1, 5]} total={4} tries={0}/>
            </store.Provider>)

    it('renders correctly', () => {
        expect(FieldTest.toJSON()).toMatchSnapshot()
    })

})