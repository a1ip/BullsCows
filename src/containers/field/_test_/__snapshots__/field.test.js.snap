// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field container renders correctly 1`] = `
<div
  className="field"
>
  <div
    className="row"
  >
    <div
      className="blocks"
    >
      <input
        autocomplete="off"
        className="input-guess"
        maxLength={1}
        name="0"
        onChange={[Function]}
        value=""
      />
      <input
        autocomplete="off"
        className="input-guess"
        maxLength={1}
        name="1"
        onChange={[Function]}
        value=""
      />
      <input
        autocomplete="off"
        className="input-guess"
        maxLength={1}
        name="2"
        onChange={[Function]}
        value=""
      />
      <input
        autocomplete="off"
        className="input-guess"
        maxLength={1}
        name="3"
        onChange={[Function]}
        value=""
      />
    </div>
    <button
      className="button"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      ‚èé
    </button>
  </div>
</div>
`;
